datafusion/sql/src/interval.rs

```diff
@@ -0,0 +1,214 @@
+                    Ok((0, 0, interval_period * SECONDS_PER_HOUR * MILLIS_PER_SECOND))
+                }
+                "minutes" | "minute" => {
+                    Ok((0, 0, interval_period * 60_f32 * MILLIS_PER_SECOND))
+                }
+                "seconds" | "second" => Ok((0, 0, interval_period * MILLIS_PER_SECOND)),
+                "milliseconds" | "millisecond" => Ok((0, 0, interval_period)),
+                _ => Err(DataFusionError::NotImplemented(format!(
+                    "Invalid input syntax for type interval: {:?}",
+                    value
+                ))),
+            }
+        };
+
+    let mut result_month: i64 = 0;
+    let mut result_days: i64 = 0;
+    let mut result_millis: i64 = 0;
+
+    let mut parts = value.split_whitespace();
+
+    loop {
+        let interval_period_str = parts.next();
+        if interval_period_str.is_none() {
+            break;
+        }
+
+        let unit = parts.next().unwrap_or(leading_field);
+
+        let (diff_month, diff_days, diff_millis) =
+            calculate_from_part(interval_period_str.unwrap(), unit)?;
+
+        result_month += diff_month as i64;
+
+        if result_month > (i32::MAX as i64) {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Interval field value out of range: {:?}",
+                value
+            )));
+        }
+
+        result_days += diff_days as i64;
+
+        if result_days > (i32::MAX as i64) {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Interval field value out of range: {:?}",
+                value
+            )));
+        }
+
+        result_millis += diff_millis as i64;
+
+        if result_millis > (i32::MAX as i64) {
+            return Err(DataFusionError::NotImplemented(format!(
+                "Interval field value out of range: {:?}",
+                value
+            )));
+        }
+    }
+
+    // Interval is tricky thing
+    // 1 day is not 24 hours because timezones, 1 year != 365/364! 30 days != 1 month
+    // The true way to store and calculate intervals is to store it as it defined
+    // It's why we there are 3 different interval types in Arrow
+    if result_month != 0 && (result_days != 0 || result_millis != 0) {
+        let result: i128 = ((result_month as i128) << 96)
+                | ((result_days as i128) << 64)
+                // IntervalMonthDayNano uses nanos, but IntervalDayTime uses milles
+                | ((result_millis * 1_000_000_i64) as i128);
+
+        return Ok(ScalarValue::IntervalMonthDayNano(Some(result)));
+    }
+
+    // Month interval
+    if result_month != 0 {
+        return Ok(ScalarValue::IntervalYearMonth(Some(result_month as i32)));
+    }
+
+    let result: i64 = (result_days << 32) | result_millis;
+    Ok(ScalarValue::IntervalDayTime(Some(result)))
+}
+
+#[cfg(test)]
+mod test {
+    use crate::assert_contains;
+
+    use super::*;
+
+    #[test]
+    fn test_parse_ym() {
```

Comment:
These tests are new

Reply:
And thank you for them! Comparing the results of differing implementations is an extremely useful test.